### DISCUSSION TITLE
Custom Environment

### DISCUSSION THREAD

**Discussion by team-daniel** [PKU-Alignment/safety-gymnasium#80] [Q&A]
I would like to make a custom environment. For example, add a goal after pressing the buttons in [Button2](https://www.safety-gymnasium.com/en/latest/environments/safe_navigation/button.html#button2). Does anyone know how I would go about this or where to look into the documentation. I have looked myself but am unable to see anything. 

Many thanks in advance!

**Comment by muchvo** [comment]
Can you explain more details about your situation? Especially the rules of your ideal environment.

**Reply by team-daniel** [reply]
Yes, thank you. Ideally, I would like a scenario such as SafetyButton0-v0 where the agent has to head to a goal after a number of button presses.

**Reply by muchvo** [reply]
Please check out [this file](https://github.com/PKU-Alignment/safety-gymnasium/blob/HEAD/safety_gymnasium/tasks/safe_navigation/button/button_level0.py). Perhaps reimplementing the two functions mentioned below can fulfill your requirements.

```python
    @property
    def goal_achieved(self):
        """Whether the goal of task is achieved."""
        for contact in self.data.contact[: self.data.ncon]:
            geom_ids = [contact.geom1, contact.geom2]
            geom_names = sorted([self.model.geom(g).name for g in geom_ids])
            # pylint: disable-next=no-member
            if any(n == f'button{self.buttons.goal_button}' for n in geom_names) and any(
                n in self.agent.body_info.geom_names for n in geom_names
            ):
                return True
        return False

    def update_world(self):
        """Build a new goal position, maybe with resampling due to hazards."""
        # pylint: disable-next=no-member
        assert self.buttons.num > 0, 'Must have at least one button.'
        self.build_goal_button()
        self.last_dist_goal = self.dist_goal()
        self.buttons.reset_timer()  # pylint: disable=no-member

```

**Reply by team-daniel** [reply]
Sorry for the late reply. I have tried that this morning and it works perfectly thank you :)
